network/framework/src/application/storage.rs

```diff
@@ -345,6 +368,61 @@ impl PeersAndMetadata {
         Ok(())
     }
 
+    fn broadcast(&self, event: ConnectionNotification) {
+        let mut listeners = self.subscribers.write();
+        let mut to_del = vec![];
+        for i in 0..listeners.len() {
+            let dest = listeners.get_mut(i).unwrap();
+            if let Err(err) = dest.try_send(event.clone()) {
+                match err {
+                    TrySendError::Full(_) => {
+                        // Tried to send to an app, but the app isn't handling its messages fast enough.
+                        // Drop message. Maybe increment a metrics counter?
+                        sample!(
+                            SampleRate::Duration(Duration::from_secs(1)),
+                            warn!("PeersAndMetadata.broadcast() failed, some app is slow"),
+                        );
+                    },
+                    TrySendError::Closed(_) => {
+                        to_del.push(i);
+                    },
+                }
+            }
+        }
+        for evict in to_del.into_iter() {
+            listeners.swap_remove(evict);
+        }
+    }
+
+    /// subscribe() returns a channel for receiving NewPeer/LostPeer events.
+    /// subscribe() immediately sends all* current connections as NewPeer events.
+    /// (* capped at NOTIFICATION_BACKLOG, currently 1000, use get_connected_peers() to be sure)
```

Comment:
nit: I'd remove this comment (I don't think we expect anywhere close to 1000 peers per node 😄 Other things would start blowing up first).